### PR TITLE
Add support for DSP 7265 Lockin

### PR DESCRIPTION
**Pull request by louisld**
Add actuator (differential measurement not supported yet) and 0D viewer for the DSP 7265 lockin.

### CODE REVIEW

**Comment by seb5g** [comment]
Hello,
are you ready for a review of your code?

**Comment by seb5g** [comment]
make sure to introduce lock mechanism as you produced both an actuator and a detector using the same ressource (controller) communicating using serial com but without any thread safe  communication protocol. Let me know if you want examples!

**Comment by louisld** [comment]
Ok I will try to implement lock mechanism.

I am interested if you have few examples.

**Comment by seb5g** [comment]
You can look at the newport plugin package:

image

image

**Comment by louisld** [comment]
I created a new class inherited from pymeasure that makes read and write functions thread safe.

File: src/pymodaq_plugins_signal_recovery/daq_move_plugins/daq_move_Lockin_DSP7265.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    """
+
+    _controller_units = 'Hz'
+    is_multiaxes = True
+    _axis_names = ['OSC']
+    _epsilon = 0.01
+    data_actuator_type = DataActuatorType.DataActuator
+
+    params = [
+        {'title': 'MultiAxes:', 'name': 'multiaxes', 'type': 'group',
```
Comment (seb5g) [inline src/pymodaq_plugins_signal_recovery/daq_move_plugins/daq_move_Lockin_DSP7265.py]: you should follow the template and not add manually these settings

File: src/pymodaq_plugins_signal_recovery/daq_move_plugins/daq_move_Lockin_DSP7265.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+         list(build_dict_from_float_list(DSP7265ThreadSafe.SENSITIVITIES, 'V'))},
+        {'title': 'Voltage (V)', 'name': 'voltage', 'type': 'float',
+         'limits': [0, 5], 'value': 1e-6},
+        {'title': 'Gain (dB)', 'name': 'gain', 'type': 'list', 'limits': GAIN}
+    ] + comon_parameters()
+
+    def ini_attributes(self) -> None:
+        self.controller: DSP7265ThreadSafe = None
+
+    def get_actuator_value(self) -> float:
```
Comment (seb5g) [inline src/pymodaq_plugins_signal_recovery/daq_move_plugins/daq_move_Lockin_DSP7265.py]: the returned type is DataActuator

File: src/pymodaq_plugins_signal_recovery/daq_move_plugins/daq_move_Lockin_DSP7265.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        -------
+        info: str
+            Id of the lockin
+        initialized: bool
+            False if initialization failed otherwise True
+        controller: object
+            Controller of the daq_move
+        """
+        try:
+            self.status.update(
```
Comment (seb5g) [inline src/pymodaq_plugins_signal_recovery/daq_move_plugins/daq_move_Lockin_DSP7265.py]: look at a recent version of the template for more concise way of doing this initialization

File: src/pymodaq_plugins_signal_recovery/daq_move_plugins/daq_move_Lockin_DSP7265.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+        Parameters
+        ----------
+        value: (float) value of the absolute target frequency
+        """
+        f = self.check_bound(f)
+        f = self.set_position_with_scaling(f)
+        self.controller.frequency = f.value()
+
+        self.target_frequency = f
```
Comment (seb5g) [inline src/pymodaq_plugins_signal_recovery/daq_move_plugins/daq_move_Lockin_DSP7265.py]: do not change current_value to current_frequency, same for target as those are used elsewhere in the parent class!!!!

File: src/pymodaq_plugins_signal_recovery/daq_move_plugins/daq_move_Lockin_DSP7265.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+    def move_rel(self, f: DataActuator) -> None:
+        """ Move the actuator to the relative target actuator value defined by
+        value
+
+        Parameters
+        ----------
+        value: (float) value of the relative target frequency
+        """
+        f = (self.check_bound(self.current_frequency + f)
```
Comment (seb5g) [inline src/pymodaq_plugins_signal_recovery/daq_move_plugins/daq_move_Lockin_DSP7265.py]: same discussion about current_frequency that should be back to current_value

File: src/pymodaq_plugins_signal_recovery/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Lockin_DSP7265.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        -------
+        info: str
+            Id of the lockin
+        initialized: bool
+            False if initialization failed otherwise True
+        controller: object
+            Controller of the daq_viewer
+        """
+        try:
+            self.status.update(
```
Comment (seb5g) [inline src/pymodaq_plugins_signal_recovery/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Lockin_DSP7265.py]: look at a recent version of the plugin template for a more concise way of doing this initialization

File: src/pymodaq_plugins_signal_recovery/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Lockin_DSP7265.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+            labels = child.value()['selected'][:]
+            subdata = [np.array([getattr(self.controller, label)])
+                       for label in labels]
+            data.append(DataFromPlugins(
+                name=child.name(),
+                data=subdata,
+                labels=labels,
+                dim='Data0D'
+            ))
+        self.data_grabed_signal.emit(data)
```
Comment (seb5g) [inline src/pymodaq_plugins_signal_recovery/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Lockin_DSP7265.py]: look at the template and use dte_signal to emit your data. These are DataToExport object containing DataFromPlugins objects. Look at the plugin template for exemples!

File: src/pymodaq_plugins_signal_recovery/hardware/dsp_7265_thread_safe.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+from pymeasure.instruments.signalrecovery import DSP7265
+
+from pymodaq.utils.logger import set_logger, get_module_name
+
+lock = Lock()
+logger = set_logger(get_module_name(__file__), add_to_console=False)
+
+class DSP7265ThreadSafe(DSP7265):
+
+    def query(self, **kwargs):
```
Comment (seb5g) [inline src/pymodaq_plugins_signal_recovery/hardware/dsp_7265_thread_safe.py]: you should probably use the read method name here to make it consistent

**Review by seb5g — CHANGES_REQUESTED** [review]
good job but quite a bit of mixed up things I highlighted. Look at a recent version of the plugin template please

**Comment by louisld** [comment]
I fixed the plugin with latest version of the template. However, I don't have access to the device right now so I need to test the change I made. I will try to do this on Friday.

**Comment by louisld** [comment]
I have an issue with multiaxes parameters in my actuator. I used `comon_parameters_fun` add the multiaxes parameters. The argument `is_multiaxes` is set to `True` but I only have one label for the axis since the actuator is one axis and the other axis are in viewer.

I saw this line inside the definition of `comon_parameter_fun`: `is_multiaxes = len(axis_names) > 1`. So that means that the argument of the function is not taken into account and my actuator is considered as one axis which cause problems for the master/slave relationship.

**Comment by louisld** [comment]
Adding a random label in `axis_names` resolve the issue and the plugin works.

**Review by seb5g — APPROVED** [review]

**Comment by seb5g** [comment]
thx for your work!

**Pull request merged**